### PR TITLE
AnyVal/value classes restrictions

### PR DESCRIPTION
**Pull request by gkossakowski**
This is merge of #1366, #1300 and some more restrictions that I collected from various value classes tickets.

Review by @odersky, @retronym.
//cc @hubertp

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 1]
```diff
@@ -1388,6 +1388,70 @@ trait Typers extends Modes with Adaptations with Tags {
...
+          }
+        }
+      for (stat <- body) {
+        def notAllowed(what: String) = unit.error(stat.pos, s"$what is not allowed in $where")
+        stat match {
+          // see https://issues.scala-lang.org/browse/SI-6444
+          // see https://issues.scala-lang.org/browse/SI-6463
+          case ClassDef(mods, _, _, _) if isValueClass =>
+            implRestriction(stat, s"nested ${ if (mods.isTrait) "trait" else "class" }")
+          case _: Import | _: ClassDef | _: TypeDef | EmptyTree => // OK
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1435]: I believe that (at least) `Block` is missing here. Test with `class X(val x: Any) extends AnyVal { { "".isEmpty; "".isEmpty } }`

Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1435]: Also `LabelDef`: `class X(val x: Any) extends AnyVal { while (true) ??? }`

Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1435]: An alternative is to remove `notAllowed("this statement")` below.

Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1435]: Oh wait, we're trying to prohibit these: "A class or trait C is ephemeral if the following holds: ... C may not have initialization statements". 

I blame it on the hour...

Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1435]: Yes, I believe that ephemeral class/trait can only have defs in their `Template` tree node so we are good with this set of conditions.

Does it make sense?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 2]
```diff
@@ -1388,6 +1388,70 @@ trait Typers extends Modes with Adaptations with Tags {
...
+          case DefDef(_, name, _, _, _, rhs) =>
+            if (stat.symbol.isAuxiliaryConstructor)
+              notAllowed("secondary constructor")
+            else if (isValueClass && (name == nme.equals_ || name == nme.hashCode_))
+              notAllowed(s"redefinition of $name method")
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I don't remember the reason for these. Is the list complete? What about `##` and `==`?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1435]: Yep, makes perfect sense.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 2 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: That's right, it's a way to ensure these sort of rewritings are safe:

https://github.com/scala/scala/blob/master/src/compiler/scala/tools/nsc/transform/PostErasure.scala#L48

I can't find the one for `hashCode` right now.

This code should document the reasons for this restriction.

Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: It's criterion number 5 from SIP:

```
C may not define equals or hashCode methods.
```

Good question about `##` and `==`. I thought that those are treated specially only for case classes, no?

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 3]
```diff
@@ -1388,6 +1388,70 @@ trait Typers extends Modes with Adaptations with Tags {
...
+            else if (stat.symbol != null && (stat.symbol hasFlag PARAMACCESSOR))
```
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I would suggest using one of the Flags helper methods. In this case case it is isParamAccessor.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 4]
```diff
@@ -1388,6 +1388,70 @@ trait Typers extends Modes with Adaptations with Tags {
...
+       * in value classes (at any nesting level).
+       *
+       * All restrictions this object imposes are probably not fundamental but require
+       * fair amount of work and testing. We are conservative for now when it comes
+       * to allowing language features to interact with value classes.
+       *  */
+      object checkEphemeralDeep extends Traverser {
+          override def traverse(tree: Tree): Unit = {
+            tree match {
+              case _: ModuleDef if isValueClass =>
```
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Shouldn't isValueClass guard be lifted?

Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Ah yes. Nested objects are disallowed both in universal traits and value classes. Nested classes are only disallowed in value classes.

I'll fix it. Thanks.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 3 above)
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: I'll fix it. Thanks.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 2 above)
Comment (harrah) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: `##` and `==` are final, so they are handled by other checks.  I think equals/hashCode should be disallowed for universal traits as well, though.  I think that is out of scope for this pull request, though.

Comment (retronym) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Overloading of these methods can also cause problems, if the optimizations are based on name rather than symbol. But it looks like that is handled correctly now: https://github.com/scala/scala/blob/master/src/compiler/scala/tools/nsc/transform/PostErasure.scala#L55

Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: @retronym: I added mention of this specific criterion in error message. Please have another look.

**Comment by gkossakowski** [comment]
All: I incorporated all of your feedback. Have another look, please.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 5]
```diff
@@ -1388,6 +1388,70 @@ trait Typers extends Modes with Adaptations with Tags {
...
+       *  */
+      object checkEphemeralDeep extends Traverser {
+          override def traverse(tree: Tree): Unit = {
+            tree match {
+              case _: ModuleDef =>
+                //see https://issues.scala-lang.org/browse/SI-6359
+                implRestriction(tree, "nested object")
+              //see https://issues.scala-lang.org/browse/SI-6444
+              // see https://issues.scala-lang.org/browse/SI-6463
+              case _: ClassDef if isValueClass =>
```
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: What I meant by saying "to lift isValueClass guard" is that it was included in every case of this Traverser

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 6]
```diff
@@ -1388,6 +1388,70 @@ trait Typers extends Modes with Adaptations with Tags {
...
     }
 
+    private def checkEphemeral(clazz: Symbol, body: List[Tree]) = {
+      // NOTE: Code appears to be messy in this method for good reason: it clearly
+      // communicates the fact that it implements rather ad-hoc, arbitrary and
+      // non-regular set of rules that identify features that interact badly with
+      // value classes. This code can be cleaned up a lot once implementation
+      // restrictions are addressed.
+      val isValueClass = !clazz.isTrait
+      def where = if (clazz.isTrait) "universal trait extending from class Any" else "value class"
```
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: You already have negation of the condition in isValueClass value.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 5 above)
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: And I guess it should be there, just put it before you match on tree.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala [hunk 7]
```diff
@@ -1395,7 +1459,7 @@ trait Typers extends Modes with Adaptations with Tags {
         unit.error(clazz.pos, "value class may not be a "+
           (if (clazz.owner.isTerm) "local class" else "member of another class"))
       if (!clazz.isPrimitiveValueClass) {
-        clazz.info.decls.toList.filter(acc => acc.isMethod && (acc hasFlag PARAMACCESSOR)) match {
+        clazz.info.decls.toList.filter(acc => acc.isMethod && acc.isParamAccessor) match {
```
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1462]: shouldn't acc.isMethod here be dropped as param accessor will always be a method?

Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1462]: Good question. This part of code was written by @odersky so I don't know what was the reasoning here. Strictly speaking `isParamAccessor` looks like this:

```
  def isParamAccessor    = hasFlag(PARAMACCESSOR)
```

so the contract we are talking about is not enforced. I hope @odersky can comment on this.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 5 above)
Comment (gkossakowski) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala]: Yes, you are right. Sorry, I got confused what is allowed and what is not in universal traits.

**Comment by gkossakowski** [comment]
@hubertp: Can you have another look? I hope I got everything right this time.

File: src/compiler/scala/tools/nsc/typechecker/Typers.scala (hunk 7 above)
Comment (hubertp) [inline src/compiler/scala/tools/nsc/typechecker/Typers.scala:1462]: Looking at the usage of isParamAccessor in Symbols it seems that my assumption was wrong. So leave it as it is.

**Comment by hubertp** [comment]
lgtm

**Comment by gkossakowski** [comment]
PLS REBUILD ALL

**Comment by gkossakowski** [comment]
@retronym: do you want to have another look before I merge? lgtm from hubert is enough for me.

**Comment by retronym** [comment]
lgtm, too.

**Comment by gkossakowski** [comment]
Cool. Thanks a lot!
